lms/djangoapps/program_enrollments/api/reading.py

```diff
@@ -377,6 +382,41 @@ def get_users_by_external_keys(program_uuid, external_user_keys):
     return users_by_external_keys
 
 
+def get_external_key_by_user_and_course(user, course_key):
+    """
+    Returns the external_user_key of the edX account/user
+    enrolled into the course
+
+    Arguments:
+        user (User):
+            The edX account representing the user in auth_user table
+        course_key (CourseKey|str):
+            The course key of the course user is enrolled in
+
+    Returns: external_user_key (str|None)
+        The external user key provided by Masters degree provider
+        Or None if cannot find edX user to Masters learner mapping
+    """
+    program_course_enrollments = ProgramCourseEnrollment.objects.filter(
+        course_enrollment__user=user,
+        course_key=course_key
+    ).order_by('-modified')
+
+    if not program_course_enrollments:
+        return None
+
+    program_enrollment = None
+    for program_course_enrollment in program_course_enrollments:
+        program_enrollment = program_course_enrollment.program_enrollment
+        if program_course_enrollment.status == ProgramCourseEnrollmentStatuses.ACTIVE:
+            # If we found this is an active program course enrollment, then according
+            # to the query order, this would be the most recent active record. we want
+            # to based on the external_user_key on this most recent active record
+            break
```

Comment:
Thanks for the order-by and the explanatory comment here; this already seems a lot more debuggable.

If there are no active records, though, this will still return the most external key of the *least* recent inactive record, which is inconsistent and potentially confusing.

How about something like this?

```python
pces = list(program_course_enrollments)
active_pces = [
    pce for pce in pces
    if pce.status == ProgramCourseEnrollmentStatuses.ACTIVE
]
try:
    # If we find an active program course enrollment, then according to
    # ..... etc.
    return next(active_pces or pces).program_enrollment.external_user_key
except StopIteration:
    return None
````

Reply:
Good suggestion. Will update

Reply:
can we order by status, then modified number, and choose the first?

Reply:
^ Ha, that's clever! It also makes enrollment-record-selection logic really easy to follow...